onboarding/mlsys/1-backprop.md

```diff
@@ -0,0 +1,414 @@
+positive, we know the minimum must be to our left, so we move in the negative
+direction (decreasing `m`).
+
+The size of our step is based on the magnitude of the slope.  If the slope is
+really steep, we proably can afford to take a really big step, but if it's
+really small we think we're getting close to the slope and we only take a small
+step.
+
+So let's define the step size `mstep` as:
+
+    mstep = - gradient * lr
+
+where `lr` is a constant multiplier, commonly called the "learning rate".
+
+But where does that `gradient` term come from?
+
+## The Gradient
+
+In the case of our simple model, we can use calculus to calculate the slope of
+our loss with respect to `m`, which for a one-variable model is also the
+gradient of the loss.
+
+We'll start by writing the loss for a single data point in terms of `m`:
+
+    loss = (ylabel - ypred)**2
+    loss = (ylabel - (m * x))**2
+
+Then apply the product rule and the chain rule from to get the derivative with
+respect to the weight `m`:
+
+    dloss/dm = - 2*x (ylabel - m*x)
+
+### Q3:
+
+Write a function to caclulate the gradient for a single data point.  It should
+take `m` and a single `(x, ylabel)` pair as inputs and return.
+
+Then write a function to calculate the total gradient for a list of
+`(x, ylabel)` pairs.
+
+Is it possible to write the second function by making multiple calls to the
+first function and combining the results?  Why or why not?
+
+## Autodifferentiation
+
+We solved our gradient symbolically because it was so simple.  But real
+gradients in real models are calculated with some magic called
+"autodifferentiation".
+
+You don't need to understand autodifferentiation, but you do need to
+understand:
+- it starts in the the "forward" pass of training a batch, and finishes in
+  the "backward" pass, and
+- it is additive.
+
+The "forward" pass is when a model makes a prediction, and the loss is
+calculated by comparing the label to the prediction.  During the forward pass,
+the deep learning library keeps information on how every weight in the model
+affects the output.  Then the "backward" pass takes the calculated loss and
+figures out how chaning each weight would affect the loss.  This output is the
+gradient.  In our toy model, the gradient is just a single number because we
+have just a single parameter (`m`), but in a model with a billion parameters the
+gradient is a billion numbers long.
+
+The additive part is important.  If you do a forward pass on a particular data
+point, follwed by a backward pass, you'll have some gradient.  If you
+immediately do another forward pass on the same data, followed by another
+background pass, you'll have twice the gradient as before.  If you do it one
+more time, you'll have three times the original gradient.
+
+### Q4:
+
+The additive nature of autodifferentiated gradients is important to
+understanding the features that Determined offers.
+
+Explain in your own words _why_ autodifferentiated gradients are additive.
```

Comment:
Maybe "The total loss is the superposition of the individual losses, so the total gradient is the superposition of the individual gradients"?

Or, "We established the gradients are additive so calculating them differently doesn't change that"?

I actually care a lot less about the answer and more about making sure that the reader understands that autodifferentiated gradients can be added together, since that's the foundation of data-parallel dtrain and gradient aggregation.

I'm also _so_ open to a better question if you have one in mind.  I don't actually like what I currently have.

Reply:
I mostly was unclear of what the question was asking.  

One possible interpretation is that it's a mathematical question: why is the addition of gradients meaningful? And I think this is your intention here.

Another interpretation is that you're asking more about implementation: why are we adding to existing gradients rather than, say, overwriting them, when we do consecutive backwards passes? Why would a user want to do multiple such backwards passes in a row without zeroing out the gradients?

Reply:
moved gradient aggregation to the end, and it doesn't have a question with it anymore.